File: calc_jte.py
import numpy as np
from Get_global_value import J_type
from Get_global_value import Ez
from Get_global_value import num_q
from f_kin_j import f_kin_j
from f_kin_e import f_kin_e
from cross import cross


def calc_jte(RR, AA, q, joints):
    n = len(joints)
    JJ_te = np.zeros((n, 3))
    if num_q == 0:
        print('Single body, there is no link')
    else:
        POS_j, ORI_j = f_kin_j(RR, AA, q, joints)
        POS_e, ORI_e = f_kin_e(RR, AA, joints)
        for i in range(n):
            A_I_i = AA[joints[i], :, :]
            if J_type[joints[i]] == 'R':
                temp = cross((A_I_i @ Ez), (POS_e - POS_j[i, :]))
                JJ_te[i, :] = temp
            else:
                JJ_te[i, :] = A_I_i @ Ez

    return JJ_te

File: src/matlab/spacedyn_v2r1/calc_jte.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% The SpaceDyn, a MATLAB toolbox for Space and Mobile Robots.
% (C)1998 The Space Robotics Lab. directed by Kazuya Yoshida,
% Tohoku University, Japan.
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%	February 4, 1998, Last modification by K.Yoshida
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%
% CALC_JTE	Translational Jacobians w.r.t.
%           for the point specified by connection vector 'joints.'
%
%   1998.1.12 A.Kurosu
%   2001.9.11 H.Hamano
%   2002.2.28 A.Noguchi modified to version up
%

function JJ_te = calc_jte(LP, SV, joints)

%global J_type Qe
%global cc ce
%global num_q 
global Ez

% Check number of joint
n = length(joints);

% If a Single body,
if LP.num_q == 0
   
   JJ_te = [];

% If a Multi body system,
else
   
   % Calculation of Joint Position
   [ POS_j, ORI_j ] = f_kin_j( LP, SV, joints );
   
   % Calculation of Effector Position
   [ POS_e, ORI_e ] = f_kin_e( LP, SV, joints );
   
   JJ_te = [];
   
   for i = 1 : 1 : n
      
      A_I_i = SV.AA(:,joints(i)*3-2:joints(i)*3);
      
      % Rotational joint
      if LP.J_type(joints(i)) == 'R'
         
         temp = cross( (A_I_i*Ez) , ( POS_e - POS_j(:,i) ) );
         JJ_te = [ JJ_te temp ];
         
      % Prismatic joint
      else
         
         JJ_te = [ JJ_te A_I_i*Ez ];
         
      end
      
   end
   
end


%%%EOF
